### PR TITLE
New announcements

### LINKED ISSUE #773: Ability for GCN Team to send Announcements and users opt in/out
- announcements email list managed like circulars list
- toggle button to subscribe/unsubscribe
- admin & moderator roles for website
- user interface for admins and moderators to send announcements

### PR DESCRIPTION
**Pull request by dakota002**
resolves #773 

Updated user email notification page:
image

Announcement email form:
image

### CODE REVIEW

File: app/routes/_gcn.user.email/email_notices.server.ts [hunk 1]
```diff
@@ -216,3 +218,25 @@ export async function sendTestEmail(to: string) {
...
+) {
+  if (!user || !user.groups.includes(moderatorGroup))
+    throw new Response(null, { status: 403 })
+
+  const db = await tables()
+  const to = (
+    await db.announcement_subscriptions.scan({
+      ProjectionExpression: 'email',
+    })
+  ).Items.map((item) => item.email)
```
Comment (lpsinger) [inline app/routes/_gcn.user.email/email_notices.server.ts:235]: You'll definitely have to paginate this. We have enough users that they won't all be returned in a single scan call. See, for example, https://github.com/nasa-gcn/gcn.nasa.gov/blob/main/app/table-streams/circulars/index.ts#L57-L65.

**Review by lpsinger — CHANGES_REQUESTED** [review]
Remember to configure the action for the unsubscribe link in https://github.com/nasa-gcn/gcn.nasa.gov/blob/main/app/routes/_gcn.unsubscribe.%24jwt/actions.server.ts#L56-L95.

**Comment by dakota002** [comment]
@lpsinger where do you think it makes sense to put links to these pages? I don't think there is a direct link to /news, and I don't have one in there yet for the form

File: app/routes/_gcn.user.email/email_notices.server.ts [hunk 2]
```diff
@@ -216,3 +218,25 @@ export async function sendTestEmail(to: string) {
...
+  await sendEmail({
```
[the remark this answers is not in the document]
Comment (dakota002) [reply app/routes/_gcn.user.email/email_notices.server.ts:236]: I'm getting this error when I test this, 
```
TypeError: Cannot read properties of undefined (reading 'template')
    at sendEmailBulk (C:\Repos\gcn.nasa.gov\app\lib\email.server.ts:101:39)
```
From the email server:
image

Are you familiar with how this is defined?

[the remark this answers is not in the document]
Comment (lpsinger) [reply app/routes/_gcn.user.email/email_notices.server.ts:236]: Yes, it's defined here:

https://github.com/nasa-gcn/gcn.nasa.gov/blob/main/src/plugins/emailOutgoing.js#L50-L57

This won't work in sandbox mode.

File: app/routes/_gcn.news.email.tsx [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+export async function action({ request }: ActionFunctionArgs) {
+  const user = await getUser(request)
+  if (!user) throw new Response(null, { status: 403 })
+  const data = await request.formData()
+  const subject = getFormDataString(data, 'subject')
+  const body = getFormDataString(data, 'body')
+  if (!subject || !body)
+    throw new Response('subject and body are required', { status: 400 })
+  await sendNewsAnnouncementEmail(subject, body, user)
+  return 'OK' //redirect('/news')
```
Comment (lpsinger) [inline app/routes/_gcn.news.email.tsx:44]: How about:

```suggestion
  return null
```

**Review by lpsinger — CHANGES_REQUESTED** [review]

**Review by lpsinger — CHANGES_REQUESTED** [review]
A better UX for after successful submission may be to show a message saying that the announcement was sent, similar to what we do in the help desk contact form:

https://github.com/nasa-gcn/gcn.nasa.gov/blob/main/app/routes/_gcn.contact.tsx#L98-L117

Sorry for flip-flopping on the pending UI.

**Comment by dakota002** [comment]
No worries, I can change that

**Comment by dakota002** [comment]
Copied a lot of the formatting from the contact page:

image

**Comment by dakota002** [comment]
@lpsinger @jracusin Is this worth including in this PR, a "News" header link:
image

**Comment by dakota002** [comment]
Also a button to link to the announcement page:

image

Is it possible to use a loader or get the user's moderator status in markdown? If not, this button is viewable to anyone, but will take them to an error page if they click it while not signed in

**Comment by lpsinger** [comment]
> Also a button to link to the announcement page:
> 
> image
> 
> Is it possible to use a loader or get the user's moderator status in markdown? If not, this button is viewable to anyone, but will take them to an error page if they click it while not signed in

That's a problem. Yes, you can have a loader in an MDX file.

**Review by lpsinger — CHANGES_REQUESTED** [review]

**Comment by dakota002** [comment]
> > Is it possible to use a loader or get the user's moderator status in markdown? If not, this button is viewable to anyone, but will take them to an error page if they click it while not signed in
> 
> That's a problem. Yes, you can have a loader in an MDX file.

Does it support useLoaderData as well? That would make things a lot easier. How do I implement that?

**Comment by lpsinger** [comment]
Yup. I don't see why not.

**Comment by dakota002** [comment]
Ah okay mdx doesn't really play well with the react hooks, but I was able to break out the button as a component and just make a new hook to use the moderator status, and it works well.

File: app/routes/_gcn/header/Header.tsx [hunk 4]
```diff
@@ -135,6 +135,14 @@ export function Header() {
...
                 Documentation
               </NavLink>,
+              <NavLink
+                className="usa-nav__link"
+                to="/news"
+                key="/news"
+                onClick={hideMobileNav}
+              >
+                News
+              </NavLink>,
```
Comment (lpsinger) [inline app/routes/_gcn/header/Header.tsx:145]: I think that I omitted News from the nav bar because it was too wide. What does this look like on desktop?

**Review by lpsinger — CHANGES_REQUESTED** [review]

**Review by lpsinger — APPROVED** [review]

**Pull request merged**